crates/red_knot_test/src/parser.rs

```diff
@@ -127,6 +129,118 @@ struct Section<'s> {
 #[newtype_index]
 struct EmbeddedFileId;
 
+#[derive(Debug)]
+pub(crate) struct CodeBlock<'s> {
+    pub(crate) code: &'s str,
+
+    /// The offset of the backticks beginning the code block within the markdown file
+    pub(crate) backtick_offset: TextSize,
+}
+
+/// Holds information about the position and the length of a single code block in a
+/// Markdown file.
+#[derive(Debug)]
+pub(crate) struct CodeBlockDimensions {
+    pub(crate) backtick_offset: TextSize,
+    pub(crate) line_count: usize,
+}
```

Comment:
Questions I'm wondering while reading through this diff for the first time:
1. I assume `CodeBlock` is meant to be an abstraction for a code block. In which case, why can't the dimensions of the code block be queried on a `CodeBlock` instance? I'm surprised there isn't a `dimensions: CodeBlockDimensions` field
2. Why is there a `backtick_offset` field on both `CodeBlock` and `CodeBlockDimensions`?

Some more docs on how these structs interrelate might be really helpful

Reply:
I removed all the overhead that was needed to make this work lazily. I hope the new structure is more clear.